### ISSUE TITLE
Broken inference between index typed object and Record

### ISSUE THREAD

**Issue by AlexGalays** [Microsoft/TypeScript#14930]
**TypeScript Version:**  2.2.2

**Code**

```ts
interface Errors {
  [field: string]: {
    key: string
    message: string
  }
}

const errors: Errors = {}

function genericObjectFunction<K extends string, V>(obj: Record<K, V>): [K, V][] {
  return []
}

/*
 Argument of type 'Errors' is not assignable to parameter of type 'Record<string, never>'.
 Index signatures are incompatible.
 Type '{ key: string; message: string; }' is not assignable to type 'never'.
 */
genericObjectFunction(errors)
```

The types seem generally compatible though, as expected this is legal:

```ts
const errorRecord: Record<string, { key: string, message: string }> = errors
```

**Expected behavior:**
`genericObjectFunction(errors)` should compile.

**Comment by aluanhaddad** [comment]
The reason this fails is that
```ts
interface Errors {
  [field: string]: {
    key: string
    message: string
  };
}
```
has no discrete keys. So the record is empty. The `Record<K, V>` type is defined by mapping keys to values and since there are no known values, the generic instantiation has no members.

Consider
```ts
const errors = { // no explicit type, inferred type has members, still compatible with interface.
  password: {
    key: 'minimumComplexity',
    message: 'Password must contain at least one non alphanumeric character'
  }
};

function genericObjectFunction<K extends string, V>(obj: Record<K, V>): [K, V][] {
  return [];
};

const genericErors = genericObjectFunction(errors);
```

**Comment by AlexGalays** [comment]
Yes, I tried the overloads as a POC and it works. It's just a bit annoying to do that on dozens of functions :)

I understand Record tries to extract information from explicit keys and values, but couldn't a simple translation exist between the two worlds?

**Comment by aluanhaddad** [comment]
> I understand Record tries to extract information from explicit keys and values, but couldn't a simple translation exist between the two worlds?

Possibly, but I feel like mixing the two could be error prone.

It is also possible that this is a bug. I do not believe it is but I could certainly be wrong.

**Comment by russelldavis** [comment]
This no longer produces an error as of v3.3.3.

Playground link: [v3.1.6, with error](https://www.typescriptlang.org/play/?ts=3.1.6#code/JYOwLgpgTgZghgYwgAgKJSgeygZ2QbwChlkBtGYCAGwBMAuZHMKUAcwF0GiSSBrCAJ4MmLEK2I8AthBw44rCMOZsJAX0LrCCTCCbJoWXA3SG8AXgKaYAVxAIwwHcgUhowBAHkARgCsI9gDFbe0cQAB4AaX0AD0gQGjwRNgAaZAA1AD4ACkxfBgAlf2waSNTMgEoGUgiy9lJ2AgkoCDBrKBAydg1CQgB6ACpiAEEoVmtpcGRMGGQwAQAHFAByE2wcJeRgPBBMMGQ4WWBWEDgvKhQwTGR5uCg4aUgoKZm5xeQlwu0oEqSxVNcAG7QDJLAB0xAAkvEINFGEcTq1mnhbihQNpJDcHGcIODkAAVBbLfDIfhCRjKMQAbmQ0lk8kU5NErGpqg2W2QOz2Bxw8NO51mV1ey0B0DBxH6vUILjcnl8-jAQTsDh0WQMa3KhCAA)
Playground link: [v3.3.3, no error](https://www.typescriptlang.org/play/?ts=3.3.3#code/JYOwLgpgTgZghgYwgAgKJSgeygZ2QbwChlkBtGYCAGwBMAuZHMKUAcwF0GiSSBrCAJ4MmLEK2I8AthBw44rCMOZsJAX0LrCCTCCbJoWXA3SG8AXgKaYAVxAIwwHcgUhowBAHkARgCsI9gDFbe0cQAB4AaX0AD0gQGjwRNgAaZAA1AD4ACkxfBgAlf2waSNTMgEoGUgiy9lJ2AgkoCDBrKBAydg1CQgB6ACpiAEEoVmtpcGRMGGQwAQAHFAByE2wcJeRgPBBMMGQ4WWBWEDgvKhQwTGR5uCg4aUgoKZm5xeQlwu0oEqSxVNcAG7QDJLAB0xAAkvEINFGEcTq1mnhbihQNpJDcHGcIODkAAVBbLfDIfhCRjKMQAbmQ0lk8kU5NErGpqg2W2QOz2Bxw8NO51mV1ey0B0DBxH6vUILjcnl8-jAQTsDh0WQMa3KhCAA)

Should this be closed?
